applications/RomApplication/custom_strategies/rom_builder_and_solver.h

```diff
@@ -256,44 +272,77 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
         }
     }
 
-    void ProjectToReducedBasis(
-        const TSystemVectorType &rX,
-        ModelPart::NodesContainerType &rNodes,
-        Vector rom_unknowns)
+
+    // Vector ProjectToReducedBasis(
+	// 	const TSystemVectorType& rX,
+	// 	ModelPart::NodesContainerType& rNodes
+	// )
+    // {
+    //     Vector rom_unknowns = ZeroVector(mRomDofs);
+    //     for(const auto& node : rNodes)
+    //     {
+    //         unsigned int node_aux_id = node.GetValue(AUX_ID);
+    //         const auto& nodal_rom_basis = node.GetValue(ROM_BASIS);
+	// 			for (int i = 0; i < mRomDofs; ++i) {
+	// 				for (int j = 0; j < mNodalDofs; ++j) {
+	// 					rom_unknowns[i] += nodal_rom_basis(j, i)*rX(node_aux_id*mNodalDofs + j);
+	// 				}
+	// 			}
+    //     }
+    //     return rom_unknowns;
+	// }
+
+    void ProjectToFineBasis(
+        const TSystemVectorType &rRomUnkowns,
+        ModelPart &rModelPart,
+        TSystemVectorType &Dx)
     {
-        for (const auto &node : rNodes){
-            unsigned int node_aux_id = node.GetValue(AUX_ID);
-            const auto &nodal_rom_basis = node.GetValue(ROM_BASIS);
-            for (int i = 0; i < mRomDofs; ++i){
-                for (int j = 0; j < mNodalDofs; ++j){
-                    rom_unknowns[i] += nodal_rom_basis(j, i) * rX(node_aux_id * mNodalDofs + j);
+        const auto dofs_begin = BaseType::mDofSet.begin();
+        const auto dofs_number = BaseType::mDofSet.size();
+
+        #pragma omp parallel firstprivate(dofs_begin, dofs_number)
+        {
+            const Matrix *current_rom_nodal_basis{nullptr};
+            unsigned int old_dof_id{};
+            #pragma omp for nowait
+            for (unsigned int k = 0; k<dofs_number; k++){
+                auto dof = dofs_begin + k;
+                if(current_rom_nodal_basis == nullptr){
+                    current_rom_nodal_basis = &(rModelPart.pGetNode(dof->Id())->GetValue(ROM_BASIS));
+                    old_dof_id = dof->Id();
+                }
+                else if(dof->Id() != old_dof_id ){

```

Comment:
i think you need to check if the dof is fixed here too

Reply:
We didnt do that before, it's passing the tests as is ...

Reply:
problem is that if you don't you may change a prescribed value... 
i am approving it here, but consider changing this in a different PR